go/vt/vtgate/planbuilder/symtab.go

```diff
@@ -139,14 +135,29 @@ func (st *symtab) Find(col *sqlparser.ColName, autoResolve bool) (rb *route, isL
 	}
 	if len(st.Colsyms) != 0 {
 		name := sqlparser.String(col)
+		var cursym *colsym
+		for _, colsym := range st.Colsyms {
+			if colsym.Alias.EqualString("*") {
+				col.Metadata = colsym
+				return colsym.Route(), true, nil
+			}
+			if colsym.Alias.EqualString(name) {
+				if cursym != nil {
+					return nil, false, fmt.Errorf("ambiguous symbol reference: %v", sqlparser.String(col))
+				}
+				cursym = colsym
+				col.Metadata = colsym
+			}
+		}
+		if cursym != nil {
+			return cursym.Route(), true, nil
+		}
 		starname := sqlparser.String(&sqlparser.ColName{
 			Name:      sqlparser.NewColIdent("*"),
 			Qualifier: col.Qualifier,
 		})
 		for _, colsym := range st.Colsyms {
-			if colsym.Alias.EqualString(name) || colsym.Alias.EqualString(starname) ||
-				colsym.Alias.EqualString("*") || colsym.ExprName.EqualString(name) ||
-				colsym.ExprName.EqualString(starname) || colsym.ExprName.EqualString("*") {
+			if colsym.QualifiedName.EqualString(name) || colsym.QualifiedName.EqualString(starname) {
 				col.Metadata = colsym
```

Comment:
This is in order to apply the rule uniformly:
* Alias will represent the unqualified symbol name for that expression. If the statement provides an explicit alias, that name will be used. Otherwise, one will be generated. If the expression is a simple column, then the base name of the column will be used as the alias.
* QualfiedName will represent the fully qualified column name, if the expression is a simple column reference. If the column expression does not reference a table (or alias), a fully qualified name will be generated based on the table alias the column references.
* Alias or QualifiedName or both could be blank if such names could not be generated. If so, those expressions cannot be referenced by other clauses of the SQL statement.

For `*` expressions, we try to apply the same rule. However, if the expression is qualified by a table name, we don't put the base `*` into this alias, because it will become a catch-all. For example, the following statement `select a.col, b.* from a join b having col2 = 1` will make us think that col2 belongs to b. We should instead return an error saying that `col2` is ambiguous.

If this explanation is clear, I'll add these as comments in the code.